[BH-1195] Enable reverse relationships

ChrisWiegman opened the pull request:
## Description

Enables reverse relationships in both publisher and in GraphQL.

https://wpengine.atlassian.net/browse/BH-1195

## Notes

There are a few differences from the mocks. First, I did not move the name field as it is not part of the field itself and would require a more significant refactor of the settings app. Second, we still only have two choices for the relationship, one-to-one and one-to-many. This allows us to enable this ticket with the existing library and will still allow for constraints such as min/max which will cover most of their intended function.

This modifies the relationship name as stored in the WordPress database. The name is used for queries and is a breaking change. The previous name was just the field name. This could cause conflicts over time as that is independent of the connection itself and multiple fields can have the same name. With reverse relationships this means that we opened the potential to return items we didn't intend to. The new name uses the field id only.

Given our current adoption I haven't done anything to update this which means existing relationships will break. I can add to this if we feel it necessary.

Finally, do notice that the reverse-slug is populated regardless of whether the two-way relationship is enabled. This is simply a UX benefit to the user and can be changed

## Testing

Appropriate PhpUnit and Codeception tests are included.

## Screenshots

![screenshot-acm lndo site-2021 09 20-12_28_43](https://user-images.githubusercontent.com/394675/134038942-5833f7ab-7755-4384-839f-7e1d0558a0a6.png)

![screenshot-acm lndo site-2021 09 20-12_27_52](https://user-images.githubusercontent.com/394675/134038983-61d7ac61-ddbd-451b-8af3-6d77ff6ea97e.png)

Commit:
Remove extra import

Commit:
Working on reverse form fields

Commit:
Add field for reverse reference

Commit:
Fix missed reference

Commit:
Properly show and hide reference fields

Commit:
Set defaults for reverse connection based on current model name

Commit:
Set field bidirectional arg as appropriate

Commit:
Ensure the register_relationship_connection method has everything it needs

Commit:
Implement reverse relationships in GraphQL

Commit:
Reverse relationship is working in GraphQL

Commit:
Fix missing ID variable that broke publisher

Commit:
Fix existing unit tests

Commit:
Add unit test that reverse relationship field is present

Commit:
Add E2E tests for configuring reverse relationship

Commit:
Use updated name and prevent editing reverse slug

Commit:
Don't disable editing reverse slug

Commit:
Add todo to update relationship names on save

Commit:
All API slugs should be readonly for editing

Commit:
Simplify function call for CodeClimate

Commit:
All relationships are saved as bidirectional

Commit:
Fix existing unit tests

Commit:
Merge commit 'dcdf929a0e1589fd80a40c199f8dcffc3714e0ca' into feature/BH-1195

Commit:
Fix e2e tests to complete and match description

nickcernis commented on includes/content-registration/custom-post-types-registration.php:
```diff
@@ -105,7 +105,8 @@ function register_relationships( $registry ) {
 
 				try {
-					$registry->define_post_to_post( $post_type, $field['reference'], $field['slug'], $args );
+					$name = $models[ $model ]['slug'] . '-' . $field['slug'] . '-' . $field['reverseSlug'];
```
Just noting that this results in a PHP warning if a relationships field was created before this PR: “Undefined index: reverseSlug”.

ChrisWiegman replied:
Good catch. I thought I had caught all those.

ChrisWiegman replied:
Fix pushed

nickcernis commented on includes/settings/js/src/components/fields/RelationshipFields.jsx:
```diff
@@ -206,6 +230,205 @@ const RelationshipFields = ({ register, data, editing, watch, errors }) => {
+								"atlas-content-modeler"
+							)}
+						</p>
+						<input
+							id="reverseSlug"
+							name="reverseSlug"
+							type="text"
+							defaultValue={
+								data?.reverseSlug
+									? data.reverseSlug
+									: toValidApiId(models[model].plural)
+							}
+							ref={register({
+								required: true,
+								maxLength: 50,
+							})}
+							onChange={(e) =>
+								onChangeGeneratedValue(e.target.value)
+							}
+							readOnly={editing}
+						/>
```
This field currently gets cleared if I:

1. Add a new relationship field.
2. Immediately click “Configure Reverse Reference” (without setting other fields such as, ”Name” and “API Identifier”).
3. Choose a model from the “Model To Reference” field.

I'm looking into the cause…

https://user-images.githubusercontent.com/647669/134879676-b614db20-f9c0-4235-853d-d189167fc64d.mp4

nickcernis replied:
I _think_ I have a fix for this in https://github.com/wpengine/atlas-content-modeler/pull/276/commits/2fcbb981fb02bb5d9b7b80c15a6661d13cd47dcb and https://github.com/wpengine/atlas-content-modeler/pull/276/commits/c6b4ed99f9b2db7bd5406d701d4bbd889e162ea8, but will do some more testing in the morning.

nickcernis commented on includes/settings/js/src/components/fields/RelationshipFields.jsx:
```diff
@@ -206,6 +230,205 @@ const RelationshipFields = ({ register, data, editing, watch, errors }) => {
+							onChange={(e) => {
+								setInputGeneratorSourceValue(e.target.value);
+								setNameCount(e.target.value.length);
+								clearErrors("reverseSlug");
```
`clearErrors` is not currently available here, so typing in the reverseName field generates errors at the moment. You can fix this by adding `clearErrors` to the `RelationshipFields` props at the top of `RelationshipFields.jsx` (it's already passed from `Form.jsx`):

```diff
const RelationshipFields = ({
	register,
	data,
	editing,
	watch,
	errors,
	setValue,
	model,
+	clearErrors,
}) => {
```

ChrisWiegman replied:
Fix pushed

Commit:
Import clearErrors

Commit:
Ensure we're always checking for the new field structure.

nickcernis commented on includes/settings/js/src/components/fields/RelationshipFields.jsx:
```diff
@@ -206,6 +231,205 @@ const RelationshipFields = ({ register, data, editing, watch, errors }) => {
+						defaultValue={
+							data?.reverseSlug
+								? data.reverseSlug
+								: toValidApiId(models[model].plural)
```
Is `toValidApiId(models[model].plural)` the correct fallback value here if the reverse reference section is not open? Or should it be an empty string (because no value has been set)?

ChrisWiegman replied:
Upon further review, yes, this is the intention as we need the slug here in a similar fashion as we talk about in your next comment.

nickcernis commented on includes/settings/js/src/components/fields/RelationshipFields.jsx:
```diff
@@ -206,6 +231,205 @@ const RelationshipFields = ({ register, data, editing, watch, errors }) => {
+							onChange={(e) =>
+								onChangeGeneratedValue(e.target.value)
+							}
+							readOnly={editing}
```
`readOnly={editing}` here means that I won't be able to set the `reverseSlug` if:

1. I create a relationship field but do not set “Configure Reverse Reference”.
2. I edit the relationship field.
3. I tick “Configure Reverse Reference”. At this point the reverseSlug is visible but not editable.

It might be that this is expected behavior, but just flagging in case we want to think about it some more.

ChrisWiegman replied:
This is expected behavior. This is the big trade-off I was worried about. TBH, in graphQL this will actually make sense

Commit:
Fix broken unit test

Commit:
Hidden fields should use value instead of defaultValue

Commit:
Use field id instead of computed name

Commit:
Fix field name in unit tests

Commit:
Revert immutability on slugs

nickcernis committed:
feature: support a default for the inputGenerator

nickcernis committed:
refactor: hide fields with CSS instead of JS

Also sets default field values using the new setDefaultInputValue
and defaultInputValue props of useInputGenerator.

Results in “hidden” fields always being rendered, so that
React Hook Form's registration process will work.
Conditional rendering was clearing the reverseSlug field during
field validation, apparently due to the initially rendered hidden
version of the field being linked to the inputGenerator, instead
of the visible version of the input field.

Intended to address the issue mentioned at
https://github.com/wpengine/atlas-content-modeler/pull/276#pullrequestreview-764026865.

nickcernis committed:
test: use classes for counter selectors

So that we don't have to worry about field order, or fields
being moved later.

Commit:
Merge commit '936f81f42651f76e9daa9dd94a3126acf5bb89ef' into feature/BH-1195

nickcernis reviewed (changes requested):
Approved but then noticed failed tests…

Commit:
Fix unit tests

Commit:
Fixing tests
